lib/Sema/MiscDiagnostics.cpp

```diff
@@ -4987,6 +4987,168 @@ static void diagUnqualifiedAccessToMethodNamedSelf(const Expr *E,
+                                              const DeclContext *DC) {
+  class DiagnoseWalker : public ASTWalker {
+    ASTContext &Ctx;
+
+  private:
+    llvm::Optional<std::string> getKeyStringValue(const Expr *keyExpr) {
+      switch (keyExpr->getKind()) {
+      case ExprKind::StringLiteral:
+      case ExprKind::FloatLiteral:
+      case ExprKind::IntegerLiteral:
+      case ExprKind::BooleanLiteral:
+      case ExprKind::NilLiteral: {
+        std::string out;
+        llvm::raw_string_ostream OS(out);
+        keyExpr->printConstExprValue(&OS, /*additionalCheck=*/nullptr);
+        return out;
+      }
+      default:
+        break;
+      }
+      return None;
+    }
+
+    std::string getKeyLiteralDescription(const Expr *keyExpr) {
+      switch (keyExpr->getKind()) {
+      case ExprKind::IntegerLiteral:
+        return "integer ";
+      case ExprKind::FloatLiteral:
+        return "floating-point ";
+      case ExprKind::BooleanLiteral:
+        return "boolean ";
+      case ExprKind::StringLiteral:
+        return "string ";
+      default:
+        break;
+      }
+      return std::string();
+    }
+
+  public:
+    DiagnoseWalker(const DeclContext *DC) : Ctx(DC->getASTContext()) {}
+
+    bool shouldWalkIntoSeparatelyCheckedClosure(ClosureExpr *expr) override {
+      return false;
+    }
+
+    bool shouldWalkIntoTapExpression() override { return false; }
+
+    std::pair<bool, Expr *> walkToExprPre(Expr *E) override {
+      if (!E || isa<ErrorExpr>(E) || !E->getType())
+        return {false, E};
+
+      const auto *DLE = dyn_cast<DictionaryExpr>(E);
+      if (!DLE)
+        return {true, E};
+
+      auto type = DLE->getType();
+      // For other types conforming with `ExpressibleByDictionaryLiteral`
+      // protocol, duplicated keys may be allowed.
+      if (!type->isDictionary()) {
+        return {true, E};
+      }
+
+      using LiteralKey = std::pair<std::string, ExprKind>;
+      using Element = std::pair<const TupleExpr *, size_t>;
+
+      std::map<LiteralKey, llvm::SmallVector<Element, 4>> groupedLiteralKeys;
+      std::map<ValueDecl *, llvm::SmallVector<Element, 4>> groupedDeclKeys;
+
+      for (size_t i = 0; i < DLE->getElements().size(); ++i) {
+        const auto *elt = DLE->getElement(i);
+        const auto *tupleElt = cast<TupleExpr>(elt);
+        const auto *keyExpr =
+            tupleElt->getElement(0)->getSemanticsProvidingExpr();
+        auto keyStringValue = getKeyStringValue(keyExpr);
+        if (keyStringValue) {
+          LiteralKey literalKey{*keyStringValue, keyExpr->getKind()};
+          groupedLiteralKeys[literalKey].push_back({tupleElt, i});
+          continue;
+        }
+
+        auto *DRE = dyn_cast<DeclRefExpr>(keyExpr);
+        if (DRE && DRE->getDecl()) {
+          auto name = DRE->getDecl()->getBaseName();
+          if (name.isSpecial())
+            continue;
+
+          groupedDeclKeys[DRE->getDecl()].push_back({tupleElt, i});
```

Comment:
Here is a problem with how you have it setup right now:

```
func test() {
  var rand: Int { get { return Int.random(in: 0...10000) } }

  let arr = [rand: "a", rand: "b"]
  print(arr)
}
```

Referencing computed property `rand` could produce different results. I'm torn between suggesting to not attempt this at all and adding more checking to this which could potentially miss some valid cases (which would result in an error in `warnings-as-errors` mode).

Reply:
Another example here would be use of property wrappers on parameters e.g. `(@Wrapper x: Int)` or declarations which control access to `wrappedValue` and `projectedValue` unbeknownst to the declaration, so we shouldn't do this probably.

Reply:
Those were a nice catch! Really didn't thought about that. Thanks

>  Referencing computed property rand could produce different results. I'm torn between suggesting to not attempt this at all and adding more checking to this which could potentially miss some valid cases (which would result in an error in warnings-as-errors mode).

Yeah, I guess we should be conservative here and do this only for literals which are really the most common case.